packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,175 @@
+  // The cache is stored in two steps; the nodes in chunks and the rest
+  // First we revive the rest, then we inject the nodes into that obj (if any)
+  // Each chunk is stored in its own file, this circumvents max buffer lengths
+  // for sites with a _lot_ of content. Since all nodes go into a Map, the order
+  // of reading them is not relevant.
+
+  const reduxCacheFolder = getReduxCacheFolder()
+
+  if (!existsSync(reduxCacheFolder)) {
+    return readFromLegacyCache()
+  }
+
+  const obj: ICachedReduxState = v8.deserialize(
+    readFileSync(reduxRestFile(reduxCacheFolder))
+  )
+
+  // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
+  const chunks = globSync(
+    reduxChunkFilePrefix(reduxCacheFolder) + `*`
+  ).map(file => v8.deserialize(readFileSync(file)))
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
+
+  if (!chunks.length) {
+    report.info(
+      `Cache exists but contains no nodes. There should be at least some nodes available so it seems the cache was corrupted. Disregarding the cache and proceeding as if there was none.`
+    )
+    // TODO: this is a DeepPartial<ICachedReduxState> but requires a big change
+    return {} as ICachedReduxState
+  }
+
+  obj.nodes = new Map(nodes)
+
+  return obj
+}
+
+function guessSafeChunkSize(values: [string, IReduxNode][]): number {
+  // Pick a few random elements and measure their size then pick a chunk size
+  // ceiling based on the worst case. Each test takes time so there's trade-off.
+  // This attempts to prevent small sites with very large pages from OOMing.
+  // This heuristic could still fail if it randomly grabs the smallest nodes.
+  // TODO: test a few nodes per each type instead of from all nodes
+
+  const nodesToTest = 11 // Very arbitrary number
+  const valueCount = values.length
+  const step = Math.max(1, Math.ceil(valueCount / nodesToTest))
+  let maxSize = 0
+  for (let i = 0; i < valueCount; i += step) {
+    const size = v8.serialize(values[i]).length
+    maxSize = Math.max(size, maxSize)
+  }
+
+  // Max size of a Buffer is 2gb (yeah, we're assuming 64bit system)
+  // https://stackoverflow.com/questions/8974375/whats-the-maximum-size-of-a-node-js-buffer
+  // Use 1.5gb as the target ceiling, allowing for some margin of error
+  return Math.floor((150 * 1024 * 1024 * 1024) / maxSize)
+}
+
+function prepareCacheFolder(
+  targetDir: string,
+  contents: ICachedReduxState
+): void {
+  // Temporarily save the nodes and remove them from the main redux store
+  // This prevents an OOM when the page nodes collectively contain to much data
+  const map = contents.nodes
+  contents.nodes = undefined
+
+  writeFileSync(reduxRestFile(targetDir), v8.serialize(contents))
+  // Now restore them on the redux store
+  contents.nodes = map
+
+  if (map) {
+    // Now store the nodes separately, chunk size determined by a heuristic
+    const values: [string, IReduxNode][] = [...map.entries()]
```

Comment:
I wonder what the memory overhead is of creating an array out of it. The cost should only be the strings because the objects aren't cloned. Is there any value on keeping the Map? (I know we have to build the "slice" function ourselves so might not be worth it)

Reply:
The map will make any lookup by id faster at runtime.

If you can think of a better way to do the slicing here I'm all ears.

One alternative would be to `map.forEach` and build up the buckets manually, but I'm convinced that would be slower due to the manual one-by-one pushes that you'd be doing a million times.

Reply:
That would be my suggestion 🤣. You're  right, let's not do this